fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java

```diff
@@ -136,6 +139,44 @@ public static Pair<List<Expression>, List<Expression>> extractExpressionForHashT
         );
     }
 
+    public static boolean hasBitmapSlot(Expression expr) {
+        return expr.getInputSlots().stream()
+                .anyMatch(slot -> slot.getDataType() == BitmapType.INSTANCE);
+    }
+
+    /**
+     * This is used for bitmap runtime filter only.
+     * Extract bitmap_contains conjunct:
+     * like: bitmap_contains(a, b) and ..., Not(bitmap_contains(a, b)) and ...,
+     * where `a` and `b` are from right child and left child, respectively.
+     *
+     * @return condition for bitmap runtime filter: bitmap_contains
+     */
+    public static List<Expression> extractBitmapRuntimeFilterConditions(List<Slot> leftSlots,
+            List<Slot> rightSlots, List<Expression> onConditions) {
+        List<Expression> result = Lists.newArrayList();
+        for (Expression expr : onConditions) {
+            BitmapContains bitmapContains = null;
+            if (expr instanceof Not) {
+                List<Expression> notChildren = ExpressionUtils.extractConjunction(expr.child(0));
+                if (notChildren.size() == 1 && notChildren.get(0) instanceof BitmapContains) {
+                    bitmapContains = (BitmapContains) notChildren.get(0);
+                }
+            } else if (expr instanceof BitmapContains) {
+                bitmapContains = (BitmapContains) expr;
+            }
+            if (bitmapContains == null) {
+                continue;
+            }
+            //first child in right, second child in left
+            if (leftSlots.containsAll(bitmapContains.child(1).collect(Slot.class::isInstance))
+                    && rightSlots.containsAll(bitmapContains.child(0).collect(Slot.class::isInstance))) {
+                result.add(expr);
+            }
```

Comment:
add a `left contains child(0) && right contains child(1)` branch is more robustness